src/webutil/tableState.js

```diff
@@ -0,0 +1,185 @@
+    sortOrder: SortOrderTypes,
+    // A key representing the sort field. It is is used to toggle sort order and increase efficiency.
+    sortName: string,
+    // A function that returns the property of the element that should be used in the sort comparator.
+    sortFn: PluckFn<T>,
+  },
+}): TableState<T> {
+  defaults = defaults || {};
+  const sortDefaults = defaults.initialSort || {};
+  const [originalArray, setOriginalArray] = useState([]);
+  const [filteredArray, setFilteredArray] = useState([]);
+  const [sortName, setSortName] = useState(sortDefaults.sortName || "");
+  const [sortOrder, setSortOrder] = useState(
+    sortDefaults.sortOrder || SortOrders.ASC
+  );
+  const [pageIndex, setPageIndex] = useState(0);
+  const [rowsPerPage, setRowsPerPageInternal] = useState(
+    defaults.initialRowsPerPage || 20
+  );
+  let _sortFn: PluckFn<T> | null = sortDefaults.sortFn || null;
+  const _filterFns: Map<string, null | ((T) => boolean)> = new Map();
+
+  const _sortAndFilter = (_array: Array<T>, _sortOrder: string) => {
+    if (_sortFn) {
+      sortBy(_array, _sortFn);
+      if (_sortOrder === SortOrders.DESC) _array.reverse();
+    }
+    setOriginalArray(_array);
+    _filter(_array);
+  };
+
+  const _filter = (_array: Array<T>) => {
+    _filterFns.forEach((fn) => {
+      if (fn) _array = _array.filter(fn);
+    });
+    setFilteredArray(_array);
+  };
+
+  const setRowsPerPage = (newRowsPerPage: number) => {
+    // Keep the user from losing their place when changing # of results per page
+    const intRowsPerPage = parseInt(newRowsPerPage, 10);
+    const currentFirstVisibleRow = pageIndex * rowsPerPage;
+    const newPage = Math.floor(currentFirstVisibleRow / intRowsPerPage);
+
+    setRowsPerPageInternal(intRowsPerPage);
+    setPageIndex(newPage);
+  };
+
+  const setSortFn = (newSortName: string, newSortFn: PluckFn<T>) => {
+    let _sortOrder = SortOrders.ASC;
+    if (newSortName === sortName) {
+      _sortOrder =
+        sortOrder === SortOrders.DESC ? SortOrders.ASC : SortOrders.DESC;
+      setFilteredArray(filteredArray.slice().reverse());
+    } else {
+      _sortFn = newSortFn;
+      _sortAndFilter(originalArray.slice(), _sortOrder);
+    }
+
+    setSortOrder(_sortOrder);
+    setSortName(newSortName);
+  };
+
+  const createOrUpdateFilterFn = (
+    key: string,
+    newFilterFn: null | ((T) => boolean)
+  ) => {
+    setPageIndex(0);
+    _filterFns.set(key, newFilterFn);
+    _filter(originalArray.slice());
+  };
+
+  const setData = (data: $ReadOnlyArray<T>) => {
+    setPageIndex(0);
+    const _array = cloneDeep(data);
+    setOriginalArray(_array);
+    _sortAndFilter(_array.slice(), sortOrder);
+  };
+
+  const currentPage = useMemo(
+    () =>
+      filteredArray.slice(
+        pageIndex * rowsPerPage,
+        pageIndex * rowsPerPage + rowsPerPage
+      ),
+    [filteredArray, pageIndex, rowsPerPage]
+  );
+
+  const length = useMemo(() => filteredArray.length, [filteredArray]);
```

Comment:
No need for useMemo here since `.length` isn't computationally intensive, and since its a plain number it will be strict equal between renders as well (e.g. you might need useMemo if you are returning an object that you want to be referentially equal between renders as a performance optimization to prevent re-renders of pure components, but its not needed here and would actually slow it down).

Reply:
I don't fully understand the nuance here because I still don't really understand React under the hood, but I'll take this suggestion and seek more understanding later.

Reply:
accepted